some unit tests for ConfigManager

alikins opened the pull request:
test the case for:
https://github.com/ansible/ansible/pull/30651
https://github.com/ansible/ansible/issues/30633

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 
 - Tests Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
test/units/config
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (config_unit_tests 6a72116a0b) last updated 2017/09/20 17:29:15 (GMT -400)
  config file = /home/adrian/src/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/adrian/src/ansible/lib/ansible
  executable location = /home/adrian/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]


```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

alikins committed:
some unit tests for ConfigManager

test the case for:
https://github.com/ansible/ansible/pull/30651
https://github.com/ansible/ansible/issues/30633

bcoca commented on test/units/config/test_manager.py:
```diff
@@ -0,0 +1,44 @@
+        self.tmp_file.close()
+        return self.tmp_file
+
+    def test_hostfile(self):
```
the tests should not really be for hostfile, as we are going to remove it.

a mock setting with 2 entries would test the functionality, not the specific setting

bcoca commented on test/units/config/test_manager.py:
```diff
@@ -0,0 +1,44 @@
+                                                 'name': 'inventory_ignore_patterns',
+                                                 'value': '/dev/null/foo,/dev/null/bar'}])
+        cm = manager.ConfigManager(conf_file=_cfg_file.name)
+        inventory_patterns = cm.data.get_setting('INVENTORY_IGNORE_PATTERNS')
```
same here, we really dont want to test the particular setting, just that both the default (dynamic template) and the nii entry are correctly parsed.

sivel commented:
I think you'll need an `__init__.py` in that dir too.

mattclay commented:
CI failure in Python 3.x unit tests:

https://app.shippable.com/github/ansible/ansible/runs/38144/4/tests

alikins commented:
@sivel wouldn't hurt to add it, but ansible-test/pytest doesn't seem to need the __init__ for test discovery

alikins commented:
Also, these tests aren't isolated from local ansible.cfg settings or environment variables.  But thats not new, none of the unit tests are (aside from a few that patch ansible.constants or the local 'C').  

ConfigManager() itself pulls from env unless it  _loop_entries or os.environ are mocked, but thats the kind of stuff having any unit tests for config are meant to find.

alikins committed:
Fix py2/py3 NamedTemporaryFile encoding diff

NamedTemporaryFile defaults to mode 'w+b' so
writing text to it via configparser fails on py3.
(And py3 configparser requires text for name/key/value)

Set tempfile mode to 'w+'

mattclay commented:
@alikins The tests should be isolated from environment variables, where are you seeing otherwise?

The tests are also supposed to be isolated from config, but a change to how config is handled broke that. I'll work on a fix.

alikins commented:
shrug

mattclay commented:
The test config isolation was fixed in https://github.com/ansible/ansible/pull/31325
